fast_llm/layers/ssm/mamba_layer.py

```diff
@@ -0,0 +1,229 @@
+import math
+from typing import Callable
+
+import torch
+from einops import rearrange, repeat
+from mamba_ssm.ops.selective_scan_interface import mamba_inner_fn, selective_scan_fn
+
+from fast_llm.engine.config_utils.tensor_space import TensorDim, TensorSpace
+from fast_llm.layers.common.linear import Linear
+from fast_llm.layers.ssm.config import SSMDimNames, SSMLayerConfig
+from fast_llm.tensor import ParameterMeta, init_ones_, kaiming_init_
+
+try:
+    from causal_conv1d import causal_conv1d_fn
+except ImportError:
+    # causal_conv1d_fn = None
+    raise ImportError("Causal conv1d not installed")
+
+
+"""
+Note: this is mostly addapted from https://github.com/Zyphra/Zamba2, similar code is aslo in https://github.com/state-spaces/mamba.
+For now it only supports training and not inference.
+This works with triton 3.1.0
+"""
+
+
+def init_A(d_state, d_inner) -> Callable[[ParameterMeta, torch.Tensor, torch.Generator], torch.Tensor]:
+    def init_(meta: ParameterMeta, tensor: torch.Tensor, generator: torch.Generator):  # noqa
+        # S4D real initialization
+        # TODO: adopt this innitialization to work for tensor parallel setting!
+        A = repeat(torch.arange(1, d_state + 1, dtype=torch.float32), "n -> d n", d=d_inner).contiguous()
+        A_log = torch.log(A)  # Keep A_log in fp32
+        if tensor.shape != A_log.shape:
+            if tensor.numel() == A_log.numel():
+                tensor_view = tensor.view(d_inner, d_state)
+                tensor_view.copy_(A_log)
+            else:
+                raise ValueError(f"Tensor size {tensor.numel()} doesn't match expected size {A_log.numel()}")
+        else:
+            tensor.copy_(A_log)
+        return tensor
+
+    return init_
+
+
+def init_dtprojbias(
+    d_inner: int, dt_max: float, dt_min: float, dt_init_floor: float, factory_kwargs: dict
+) -> Callable[[ParameterMeta, torch.Tensor, torch.Generator], torch.Tensor]:
+    def init_(meta: ParameterMeta, tensor: torch.Tensor, generator: torch.Generator):  # noqa
+        dt = torch.exp(
+            torch.rand(d_inner, **factory_kwargs) * (math.log(dt_max) - math.log(dt_min)) + math.log(dt_min)
+        ).clamp(min=dt_init_floor)
+        # Inverse of softplus: https://github.com/pytorch/pytorch/issues/72759
+        inv_dt = dt + torch.log(-torch.expm1(-dt))
+        tensor.copy_(inv_dt)
+        return tensor
+
+    return init_
+
+
+class MambaLayer(torch.nn.Module):
```

Comment:
This doesn't work with TP, need to explicitly prevent. (Not sure about PP).

Reply:
we're gonna have TP eventually, not in scope for this PR though